core/jvm/src/test/scala/zio/IOSpec.scala

```diff
@@ -579,4 +583,39 @@ class IOSpec(implicit ee: org.specs2.concurrent.ExecutionEnv) extends TestRuntim
 
     goodCase and partialBadCase and badCase
   }
+
+  def testForeachParN_Threads = {
+    val n   = 10L
+    val seq = 0 to 100000
+    val res = unsafeRun(IO.foreachParN(n)(seq)(UIO.succeed))
+    res must be_===(seq)
+  }
+
+  def testForeachParN_Parallel = {
+    val io = for {
+      p <- Promise.make[Nothing, Unit]
+      _ <- UIO.foreachParN(2)(List(UIO.never, p.succeed(())))(a => a).fork
+      _ <- p.await
+    } yield true
+    unsafeRun(io)
+  }
+
+  def testForeachParN_Error = {
+    val ints = List(1, 2, 3, 4, 5, 6)
+    val odds = ZIO.foreachParN(4)(ints) { n =>
+      if (n % 2 != 0) ZIO.succeed(n) else ZIO.fail("not odd")
+    }
+    unsafeRun(odds.either) must_=== Left("not odd")
+  }
+
+  def testForeachParN_Interruption = {
+    import zio.duration._
+    val actions = List(
+      clock.sleep(1000.millis) *> ZIO.fail("A"),
```

Comment:
This takes a whole second. Can we reduce the timeout, or use a latch? Test speed is critical as the test suite grows forever without bound.

Reply:
I refactored this test to avoid using timeouts.